pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/OpChainSchedulerService.java

```diff
@@ -31,151 +33,84 @@
-                returnedErrorBlock = true;
-                LOGGER.error("({}): Completed erroneously {} {}", operatorChain, operatorChain.getStats(),
-                    result.getDataBlock().getExceptions());
-              } else {
-                LOGGER.debug("({}): Completed {}", operatorChain, operatorChain.getStats());
-              }
+              LOGGER.debug("({}): Completed {}", operatorChain, operatorChain.getStats());
             }
           } catch (Exception e) {
             LOGGER.error("({}): Failed to execute operator chain! {}", operatorChain, operatorChain.getStats(), e);
             thrown = e;
           } finally {
-            if (returnedErrorBlock || thrown != null) {
+            if (returnedErrorBlock != null || thrown != null) {
+              if (thrown == null) {
+                String blockMsg;
+                if (returnedErrorBlock.getDataBlock() instanceof MetadataBlock) {
+                  MetadataBlock metadataBlock = (MetadataBlock) returnedErrorBlock.getDataBlock();
+                  blockMsg = String.join(", ", metadataBlock.getExceptions().values());
+                } else {
+                  blockMsg = "Unknown";
+                }
+                thrown = new RuntimeException("Error block " + blockMsg);
+              }
               cancelOpChain(operatorChain, thrown);
             } else if (isFinished) {
               closeOpChain(operatorChain);
             }
           }
         }
       });
-    }
+    _submittedOpChainMap.put(operatorChain.getId(), scheduledFuture);
   }
 
-  /**
-   * Register a new operator chain with the scheduler.
-   *
-   * @param operatorChain the chain to register
-   */
-  public final void register(OpChain operatorChain) {
-    operatorChain.getStats().queued();
-    _scheduler.register(operatorChain);
-    LOGGER.debug("({}): Scheduler is now handling operator chain listening to mailboxes {}. "
-            + "There are a total of {} chains awaiting execution.", operatorChain,
-        operatorChain.getReceivingMailboxIds(),
-        _scheduler.size());
-  }
-
-  /**
-   * Async cancel a request. Request will not be fully cancelled until the next time opChain is being polled.
-   *
-   * @param requestId requestId to be cancelled.
-   */
-  public final void cancel(long requestId) {
-    _cancelledRequests.put(requestId, requestId);
-  }
-
-  /**
-   * This method should be called whenever data is available for an {@link OpChain} to consume.
-   * Implementations of this method should be idempotent, it may be called in the scenario that no data is available.
-   *
-   * @param opChainId the identifier of the operator chain
-   */
-  public final void onDataAvailable(OpChainId opChainId) {
-    _scheduler.onDataAvailable(opChainId);
-  }
-
-  // TODO: remove this method after we pipe down the proper executor pool to the v1 engine
-  public ExecutorService getWorkerPool() {
-    return _workerPool;
+  @Override
+  public void cancel(long requestId) {
```

Comment:
We probably want a different API to directly cancel an `OpChain` (or `OpChainId`) because that is much cheaper than cancelling a request id

Reply:
I may not have the whole picture but this way to cancel a request doesn't seem so expensive to me TBH. I would not expect to receive thousands of cancellation request per second, but a few per day. The cost of cancelling them is also not that high. Is there something I'm not getting?

Reply:
My concern is more on the automatic cancellation which happens when `QueryRunner.processQuery()` throws exception (this can potentially happen a lot of times). Seems when the exception happens, the current `OpChain` is never registered, so is it intentional to cancel other submitted OpChains? Since we will always get the cancellation from the coordinator in such scenario, we can also just rely on the upstream cancellation

Reply:
This code came from @walterddr's PR, so TBH I don't have the context to answer the question.

Anyway, is this a blocking issue? The plan was to have this merged before the end of the week. Can we do that with this cancellation mechanism or should we try to find another way?

Reply:
yes we rely on upstream cancellation

Reply:
but we should still use a timeout cache or something as a final fallback. b/c the grpc cancel signal can also get lost in the wire (we don't retry cancel or monitor cancel grpc call status)

Reply:
I don't get that. In case the cancel signal is lost, what would happen is that the node that should have receive it would execute the query as it was expected. I assume that the receiving mailbox will be eventually closed but even if it doesn't the reader would timeout. That means that the runnable will throw an exception that will be caught here and therefore cancel the op chain.

By adding a timeout here we would add two competing timeouts (the one reading the mailbox and the one here) which would make the code more undeterministic without giving us more guarantees.

Also, it doesn't seem to be what concerns Jackie. He seems to be concerned about the cost of calling this method several times per query.